python_modules/libraries/dagster-openai/dagster_openai/resources.py

```diff
@@ -0,0 +1,254 @@
+from collections import defaultdict
+from contextlib import contextmanager
+from enum import Enum
+from functools import wraps
+from typing import Generator, List, Optional
+from weakref import WeakKeyDictionary
+
+from dagster import AssetExecutionContext, ConfigurableResource, InitResourceContext
+from openai import Client
+from pydantic import Field, PrivateAttr
+
+
+class ApiResourceEnum(Enum):
+    """Legacy and current resources of the OpenAI API v1."""
+
+    COMPLETIONS = "completions"
+    CHAT = "chat"
+    EMBEDDINGS = "embeddings"
+    FILES = "files"
+    IMAGES = "images"
+    AUDIO = "audio"
+    MODERATIONS = "moderations"
+    MODELS = "models"
+    FINE_TUNING = "fine_tuning"
+
+
+API_RESOURCES_TO_ENDPOINT_METHODS_MAPPING = {
+    "completions": [["create"]],
+    "chat": [["completions", "create"]],
+    "embeddings": [["create"]],
+    "files": [
+        ["create"],
+        ["retrieve"],
+        ["list"],
+        ["delete"],
+        ["content"],
+        ["retrieve_content"],
+        ["wait_for_processing"],
+    ],
+    "images": [["create_variation"], ["edit"], ["generate"]],
+    "audio": [["transcriptions", "create"], ["translations", "create"], ["speech", "create"]],
+    "moderations": [["create"]],
+    "models": [["retrieve"], ["list"], ["delete"]],
+    "fine_tuning": [
+        ["jobs", "create"],
+        ["jobs", "retrieve"],
+        ["jobs", "list"],
+        ["jobs", "cancel"],
+        ["jobs", "list_events"],
+    ],
+}
+
+context_to_counters = WeakKeyDictionary()
+
+
+def add_to_asset_metadata(
```

Comment:
re: [1], then you probably change it so it's update dict instead of add single value to a metadata key